ta4j-core/src/main/java/org/ta4j/core/util/BarAggregator.java

```diff
@@ -0,0 +1,110 @@
+
+	/**
+	 * Aggregates the bars of the <code>barSeries</code> by
+	 * <code>timePeriod</code>. The new <code>timePeriod</code> must be a
+	 * multiple of the actual time period.
+	 * 
+	 * @param barSeries the actual TimeSeries
+	 * @param timePeriod the actual timePeriod
+	 * @return the aggregated TimeSeries
+	 */
+	public static TimeSeries aggregateBarSeries(TimeSeries barSeries, Duration timePeriod) {
+		String name = barSeries.getName() + "_upscaled_to_" + timePeriod;
+		return aggregateBarSeries(name, barSeries, timePeriod);
+	}
+
+	/**
+	 * Aggregates the bars of the <code>barSeries</code> by
+	 * <code>timePeriod</code>. The new <code>timePeriod</code> must be a
+	 * multiple of the actual time period.
+	 * 
+	 * @param name       the name of the returned TimeSeries
+	 * @param barSeries the actual TimeSeries
+	 * @param timePeriod the actual timePeriod
+	 * @return the aggregated TimeSeries
+	 */
+	public static TimeSeries aggregateBarSeries(String name, TimeSeries barSeries, Duration timePeriod) {
+		List<Bar> sumBars = aggregateBars(barSeries.getBarData(), timePeriod);
+		return new BaseTimeSeries(name, sumBars);
+	}
+
+	/**
+	 * Aggregates a list of bars by <code>timePeriod</code>.The new
+	 * <code>timePeriod</code> must be a multiple of the actual time period.
+	 * 
+	 * @param bars       the actual bars
+	 * @param timePeriod the new timePeriod
+	 * @return the aggregated bars with new <code>timePeriod</code>
+	 */
+	public static List<Bar> aggregateBars(List<Bar> bars, Duration timePeriod) {
+		List<Bar> sumBars = new ArrayList<>();
+		if (bars.isEmpty()) return sumBars;
+		// get the actual time period
+		Duration actualDur = bars.iterator().next().getTimePeriod();
+		// check if new timePeriod is a multiple of actual time period
+		boolean isNotMultiple = timePeriod.getSeconds() % actualDur.getSeconds() != 0;
+
+		if (isNotMultiple) {
+			throw new IllegalArgumentException("Cannot aggregate bars: the new timePeriod must be a multiple of the actual timePeriod.");
+		}
+
+		int i = 0;
+		final Num zero = bars.iterator().next().getOpenPrice().numOf(0);
+		while (i < bars.size()) {
+			Bar b1 = bars.get(i);
+			ZonedDateTime beginTime = b1.getBeginTime();
+			Num open = b1.getOpenPrice();
+			Num high = b1.getHighPrice();
+			Num low = b1.getLowPrice();
+
+			// set to ZERO
+			Num close = zero;
+			Num volume = zero;
+			Num amount = zero;
+			Duration sumDur = Duration.ZERO;
+
+			while (sumDur.compareTo(timePeriod) < 0) {
+				if (i < bars.size()) {
+					Bar b2 = bars.get(i);
+
+					if (b2.getHighPrice().isGreaterThan(high)) {
+						high = b2.getHighPrice();
+					}
+					if (b2.getLowPrice().isLessThan(low)) {
+						low = b2.getLowPrice();
+					}
+					close = b2.getClosePrice();
+					volume = volume.plus(b2.getVolume());
+					amount = amount.plus(b2.getAmount());
+				}
+				sumDur = sumDur.plus(actualDur);
+				i++;
+			}
+
+			// add only bars with elapsed timePeriod
+			if (i <= bars.size()) {
```

Comment:
This solution is not flexible, it should be configurable if only bars with elapsed timePeriod are allowed

Reply:
good idea. put latest unclosed bar with close = null.

Reply:
No, it does not mean that is close == null. 

See DurationBarAggregatorTest. upscaledTo10DayBarsNotOnlyFinalBars in https://github.com/ta4j/ta4j/pull/421